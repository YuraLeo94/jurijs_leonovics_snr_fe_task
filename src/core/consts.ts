export const BALANCE_PATH = "/balances";

const PATH = {
  HOME: "/",
  BALANCE: "/balances",
  LOGIN: "/login",
};

const NAV_MENU = {
  HOME: "Home",
  BALANCE: "Balances Tables",
  LOGIN: "Login",
};

const BALANCE_TABLE = {
  TABLE: {
    COL1: "Name",
    COL2: "Balance",
    COL3: "Delete",
  },
  BUTTONS: {
    INCREMENT: "+ Column",
    DECREMENT: "- Column",
  },
};

const AUTH = {
  LOGIN: {
    EMAIL_ERROR_MESSAGE: "Incorrect email",
    EMAIL_PLACEHOLDER: "Email",
    PASSWORD_ERROR_MESSAGE: "Incorrect password",
    PASSWORD_PLACEHOLDER: "Password",
    OTP_FORM_TITLE: "Enter OTP",
    OTP_FORM_BUTTON_LABEL: "Submit",
    OTP_PLACEHOLDER: "Email",
    LOGIN_FORM_TITLE: "Login",
    LOGIN_FORM_BUTTON_LABEL: "Next",
  },
};

const FETCH_DATA_GENERIC_ERROR_MESSAGE =
  "Failed to fetch data. Please try again later.";
const LOADING_MESSAGE = "...Loading";

const COLUMNS_DEFAULT_COUNT = 3;

export {
  FETCH_DATA_GENERIC_ERROR_MESSAGE,
  COLUMNS_DEFAULT_COUNT,
  BALANCE_TABLE,
  NAV_MENU,
  PATH,
  LOADING_MESSAGE,
  AUTH,
};
