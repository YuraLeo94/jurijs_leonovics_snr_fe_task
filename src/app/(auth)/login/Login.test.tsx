import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import Login from "./page";
import { useRouter } from "next/navigation";
import { AUTH, BALANCE_PATH } from "@/core/consts";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

describe("Login Component", () => {
  let mockPush: jest.Mock;

  beforeEach(() => {
    mockPush = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
  });

  test("shows error for incorrect email", () => {
    render(<Login />);

    const emailInput = screen.getByPlaceholderText(
      AUTH.LOGIN.EMAIL_PLACEHOLDER
    );
    const passwordInput = screen.getByPlaceholderText(
      AUTH.LOGIN.PASSWORD_PLACEHOLDER
    );
    const loginButton = screen.getByText(AUTH.LOGIN.LOGIN_FORM_BUTTON_LABEL);

    fireEvent.change(emailInput, { target: { value: "incorrect@email.com" } });
    fireEvent.change(passwordInput, { target: { value: "correct-password" } });
    fireEvent.click(loginButton);

    expect(
      screen.getByText(AUTH.LOGIN.EMAIL_ERROR_MESSAGE)
    ).toBeInTheDocument();
  });

  test("shows error for incorrect password", () => {
    render(<Login />);

    const emailInput = screen.getByPlaceholderText(
      AUTH.LOGIN.EMAIL_PLACEHOLDER
    );
    const passwordInput = screen.getByPlaceholderText(
      AUTH.LOGIN.PASSWORD_PLACEHOLDER
    );
    const loginButton = screen.getByText(AUTH.LOGIN.LOGIN_FORM_BUTTON_LABEL);

    fireEvent.change(emailInput, { target: { value: "correct@email.com" } });
    fireEvent.change(passwordInput, {
      target: { value: "incorrect-password" },
    });
    fireEvent.click(loginButton);

    expect(
      screen.getByText(AUTH.LOGIN.PASSWORD_ERROR_MESSAGE)
    ).toBeInTheDocument();
  });

  test("navigates to OTP step on correct email and password", async () => {
    render(<Login />);

    const emailInput = screen.getByPlaceholderText(
      AUTH.LOGIN.EMAIL_PLACEHOLDER
    );
    const passwordInput = screen.getByPlaceholderText(
      AUTH.LOGIN.PASSWORD_PLACEHOLDER
    );
    const loginButton = screen.getByText(AUTH.LOGIN.LOGIN_FORM_BUTTON_LABEL);

    fireEvent.change(emailInput, { target: { value: "correct@email.com" } });
    fireEvent.change(passwordInput, { target: { value: "correct-password" } });
    fireEvent.click(loginButton);

    await waitFor(
      () =>
        expect(screen.getByText(AUTH.LOGIN.OTP_FORM_TITLE)).toBeInTheDocument(),
      {
        timeout: 2000,
      }
    );
  });

  test("navigates to BALANCE_PATH after OTP submit", async () => {
    render(<Login />);

    const emailInput = screen.getByPlaceholderText(
      AUTH.LOGIN.EMAIL_PLACEHOLDER
    );
    const passwordInput = screen.getByPlaceholderText(
      AUTH.LOGIN.PASSWORD_PLACEHOLDER
    );
    const loginButton = screen.getByText(AUTH.LOGIN.LOGIN_FORM_BUTTON_LABEL);

    fireEvent.change(emailInput, { target: { value: "correct@email.com" } });
    fireEvent.change(passwordInput, { target: { value: "correct-password" } });
    fireEvent.click(loginButton);

    await waitFor(
      () =>
        expect(screen.getByText(AUTH.LOGIN.OTP_FORM_TITLE)).toBeInTheDocument(),
      {
        timeout: 2000,
      }
    );

    const otpInput = screen.getByPlaceholderText(AUTH.LOGIN.OTP_PLACEHOLDER);
    fireEvent.change(otpInput, { target: { value: "123456" } });
    const submitButton = screen.getByText(AUTH.LOGIN.OTP_FORM_BUTTON_LABEL);

    fireEvent.click(submitButton);

    await waitFor(() => expect(mockPush).toHaveBeenCalledWith(BALANCE_PATH), {
      timeout: 2000,
    });
  });

  test("matches the snapshot", () => {
    const { container } = render(<Login />);
    expect(container).toMatchSnapshot();
  });
});
