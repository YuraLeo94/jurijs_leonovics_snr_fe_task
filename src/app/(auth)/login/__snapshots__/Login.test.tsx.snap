// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches the snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center min-h-screen bg-gray-100"
  >
    <div
      class="p-6 bg-white rounded-lg shadow-md w-96 relative"
    >
      <h2
        class="text-xl font-bold mb-4"
      >
        Login
      </h2>
      <input
        class="border p-2 rounded w-full mb-2"
        placeholder="Email"
        type="email"
        value=""
      />
      <input
        class="border p-2 rounded w-full mb-2"
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        class="bg-blue-500 text-white p-2 mt-2 rounded w-full hover:bg-blue-700"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
